templates/preferences.html

```diff
@@ -0,0 +1,210 @@
+                Add Row
+            </button>
+        </form>
+    </div>
+{% endblock %}
+
+{% block additionalfooter %}
+    <script>
+        $(document).ready(function () {
+            // Handle year select change
+            handleYearSelectChange();
+
+            //Handle preferences
+            $('#addPreferencesButton').click(function () {
+                $('#staticPreferencesTable').hide();
+                $('#editablePreferencesTable').show();
+                $('#addPreferencesButton').hide();
+                $('#savePreferencesButton').show();
+                $('#cancelEditButton').show();
+                $('#addRowButton').show();
+            });
+
+            $('#cancelEditButton').click(function () {
+                $('#editablePreferencesTable').hide();
+                $('#staticPreferencesTable').show();
+                $('#addPreferencesButton').show();
+                $('#savePreferencesButton').hide();
+                $('#cancelEditButton').hide();
+                $('#addRowButton').hide();
+            });
+
+            $('#addRowButton').click(function () {
+                const newRow = $('#template-row').clone().removeAttr('id').removeAttr('style');
+                $('#editablePreferencesTable tbody').append(newRow);
+
+                $("#editablePreferencesTable tbody").sortable({
+                    items: 'tr',
+                    handle: '.move-icon',
+                    axis: 'y',
+                    cursor: 'move'
+                });
+            });
+
+            // Event listener pour le bouton de suppression
+            $(document).on('click', '.delete-icon', function () {
+                $(this).closest('tr').remove();
+            });
+
+            $('#savePreferencesButton').click(function () {
+                const preferences = [];
+                !$('#editablePreferencesTable tbody tr').not('#template-row').each(function () {
+                    const courseValue = $(this).find('select.course-select').val();
+
+                    if (courseValue) {
+                        const [course_id, course_year] = courseValue.split("-");
+                        preferences.push({course_id: course_id, course_year: course_year});
+                    }
+                });
+
+                const data = {
+                    preferences: preferences,
+                };
+
+                $.ajax({
+                    type: 'POST',
+                    url: "{{ url_for('course_preference.save_preference') }}",
+                    contentType: 'application/json',
+                    data: JSON.stringify(data),
+                    success: function (response) {
+                        alert('Preferences updated successfully.');
+                        location.reload(); // Refresh the page to show the updated preferences
+                    },
+                    error: function (error) {
+                        alert('Error updating the preferences.');
```

Comment:
Here an error message will be shown only for the unmanaged errors in the python code, that is, when the app crashes.